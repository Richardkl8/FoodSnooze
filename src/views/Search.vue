<template>
    <div class="container">
        <section id="searched-items ">
            <custom-h2 class="mb-4 mt-5">
                <b>You searched for:</b>
                {{this.$route.params.query}}
            </custom-h2>
            <div v-if="foundRecipes.length > 0" class="row my-5">
                <div v-for="recipe in foundRecipes"
                     :key="recipe.mealId"
                     class="col-xs-12 col-sm-6 col-md-4 col-lg-3 ">
                    <recipe-card
                        :recipe="recipe"
                        @click.native="$_goToRecipePage(relatedMeals[index].idMeal)">
                    </recipe-card>
                </div>
            </div>
            <div v-else>
                <div class="row">
                    <div class="col">
                        <img src="https://images.assetsdelivery.com/compings_v2/lineartist/lineartist1905/lineartist190500646.jpg" alt="crying chef">
                        <custom-h3><b>Sorry, </b>we couldn't find any recipes for: {{this.$route.params.query}} </custom-h3>
                    </div>
                </div>
            </div>
        </section>
    </div>
</template>

<script>
    import RecipeCard from '@/components/RecipeCard.vue';
    import goToPageMixin from '../mixins/goToPageMixin';

    export default {
        name: 'Home',
        components: {
            RecipeCard,
        },
        mixins: [
            goToPageMixin,
        ],
        inject: [
            'apiService',
        ],
        watch: {
            $route() {
                this.getRecipeByQuery(this.$route.params.query);
            },
        },
        data() {
            return {
                foundRecipes: [],
            };
        },
        methods: {
            getRecipeByQuery() {
                this.apiService.getRecipeByQuery(this.$route.params.query)
                    .then((response) => {
                        if (response) {
                            this.foundRecipes = response;
                        } else {
                            this.foundRecipes = [];
                        }
                    });
            },
        },
        created() {
            this.getRecipeByQuery();
        },
    };
</script>

<style lang="scss" scoped>
    .recipe-image {
        max-width: 100%;
    }

    ul {
        padding: 0;
        margin: 0;
        list-style-type: none;
    }

    li {
        margin: 1rem 0;
    }

    .ingredients {
        padding: 3rem 2rem;
        border: solid 2px #2DB396;
        color: #fff;
    }
</style>
