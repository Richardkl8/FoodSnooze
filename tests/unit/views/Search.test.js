import { shallowMount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Search from '../../../src/views/Search.vue';
import stubs from '../stubs/stubs';


const responseGetRecipeByQuery = [
    {
        idMeal: '12',
        strMealThumb: 'image.jpg',
        strMeal: 'tuna',
        strCategory: 'fish',
    },
    {
        idMeal: '12',
        strMealThumb: 'image.jpg',
        strMeal: 'tuna',
        strCategory: 'fish',
    },
    {
        idMeal: '12',
        strMealThumb: 'image.jpg',
        strMeal: 'tuna',
        strCategory: 'fish',
    },
];

const provide = {
    apiService: {
        getRecipeByQuery: jest.fn().mockResolvedValue(responseGetRecipeByQuery),
    },
};

const mocks = {
    $route: {
        params: {
            query: 'fish',
        },
        path: '/search/fish',
    },
};


describe('Search page', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Search, {
            provide,
            stubs: stubs.general,
            mocks,
        });
    });

    test('Should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Should call getRecipeByQuery with route params when component is created', () => {
        expect(wrapper.vm.apiService.getRecipeByQuery).toHaveBeenCalledWith('fish');
    });

    test('Should show error when search query returns empty array', async () => {
       expect(wrapper.find('[testId="recipe-card-search"]').isVisible()).toBe(true);
       wrapper.setData({ foundRecipes: [] });
       await flushPromises();
       expect(wrapper.find('[testId="recipe-card-search"]').exists()).toBe(false);
    });

    test('Should set foundRecipes to an empty array when response is null', async () => {
        provide.apiService.getRecipeByQuery.mockResolvedValue(null);
        wrapper = shallowMount(Search, {
            provide,
            stubs: stubs.general,
            mocks,
        });
        wrapper.vm.apiService.getRecipeByQuery();
        await flushPromises();
        expect(wrapper.vm.$data.foundRecipes).toEqual([]);
    });
});
