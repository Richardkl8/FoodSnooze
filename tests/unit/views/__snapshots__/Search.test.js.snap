// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search page Should match snapshot 1`] = `
<div class="container">
  <section id="searched-items ">
    <custom-h2-stub class="mb-4 mt-5"><b>Search results for:</b>
      fish
    </custom-h2-stub>
    <div class="row my-5">
      <div class="col-xs-12 col-md-4">
        <router-link-stub to="/recipe/12">
          <recipe-card-stub recipe="[object Object]" testid="recipe-card-search"></recipe-card-stub>
        </router-link-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <router-link-stub to="/recipe/12">
          <recipe-card-stub recipe="[object Object]" testid="recipe-card-search"></recipe-card-stub>
        </router-link-stub>
      </div>
      <div class="col-xs-12 col-md-4">
        <router-link-stub to="/recipe/12">
          <recipe-card-stub recipe="[object Object]" testid="recipe-card-search"></recipe-card-stub>
        </router-link-stub>
      </div>
    </div>
  </section>
</div>
`;
